hashring/hashring.go

```diff
@@ -66,119 +68,77 @@ func (r *HashRing) RegisterListener(l events.EventListener) {
 	r.listeners = append(r.listeners, l)
 }
 
-func NewHashRing(hashfunc func([]byte) uint32, replicaPoints int) *HashRing {
-	ring := &HashRing{
-		hashfunc:      hashfunc,
+// New instantiates and returns a new HashRing.
+func New(hashfunc func([]byte) uint32, replicaPoints int) *HashRing {
+	r := &HashRing{
 		replicaPoints: replicaPoints,
+		hashfunc: func(str string) int {
+			return int(hashfunc([]byte(str)))
+		},
 	}
 
-	ring.servers.byAddress = make(map[string]bool)
-	ring.servers.tree = &RedBlackTree{}
-
-	return ring
-}
-
-func (r *HashRing) Checksum() uint32 {
-	r.servers.RLock()
-	checksum := r.servers.checksum
-	r.servers.RUnlock()
-
-	return checksum
+	r.servers = make(map[string]struct{})
+	r.tree = &RedBlackTree{}
+	return r
 }
 
-// computeChecksum computes checksum of all servers in the ring
-func (r *HashRing) computeChecksum() {
-	var addresses sort.StringSlice
-	var buffer bytes.Buffer
-
-	r.servers.Lock()
-	for address := range r.servers.byAddress {
-		addresses = append(addresses, address)
-	}
-
-	addresses.Sort()
-
-	for _, address := range addresses {
-		buffer.WriteString(address)
-		buffer.WriteString(";")
-	}
-
-	old := r.servers.checksum
-	r.servers.checksum = farm.Fingerprint32(buffer.Bytes())
-	r.emit(events.RingChecksumEvent{
-		OldChecksum: old,
-		NewChecksum: r.servers.checksum,
-	})
-
-	r.servers.Unlock()
-}
-
-func (r *HashRing) AddServer(address string) {
-	if r.HasServer(address) {
-		return
+// Lookup returns the owner of the given key.
+func (r *HashRing) Lookup(key string) (string, bool) {
+	strs := r.LookupN(key, 1)
+	if len(strs) == 0 {
+		return "", false
 	}
-
-	r.addReplicas(address)
-	r.emit(events.RingChangedEvent{ServersAdded: []string{address}})
-	r.computeChecksum()
+	return strs[0], true
 }
 
-// inserts server replicas into ring
-func (r *HashRing) addReplicas(server string) {
-	r.servers.Lock()
-	r.servers.byAddress[server] = true
+// LookupN returns the N servers that own the given key. Duplicates in the form
+// of virtual nodes are skipped to maintain a list of unique servers. If there
+// are less servers then N, we simply return all existing servers.
+func (r *HashRing) LookupN(key string, n int) []string {
```

Comment:
Why reorder all the functions?  There may actually be a good reason (not asserting otherwise), but it makes the diff harder to follow so should be well-motivated.